File: java/quantile-estimator/src/main/java/org/mydotey/quantile/classic/ClassicQuantileEstimator.java
package org.mydotey.quantile.classic;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import org.mydotey.quantile.QuantileEstimator;
import org.mydotey.quantile.value.Calculator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author koqizhao
 *
 * Apr 1, 2018
 */
public class ClassicQuantileEstimator<T> implements QuantileEstimator<T> {

    private static Logger _logger = LoggerFactory.getLogger(ClassicQuantileEstimator.class);

    private ClassicQuantileEstimatorConfig<T> _config;

    private volatile List<T> _values;

    public ClassicQuantileEstimator(ClassicQuantileEstimatorConfig<T> config) {
        Objects.requireNonNull(config, "config is null");
        _config = config;

        _values = new LinkedList<>();
    }

    @Override
    public void add(T value) {
        _values.add(value);
    }

    @Override
    public Map<Double, T> get(List<Double> quantiles) {
        if (_values.isEmpty())
            return null;

        _values.sort(_config.getComparator());

        HashMap<Double, T> results = new HashMap<>();
        double n = _values.size() - 1;
        Calculator<T> caculator = _config.getCalculator();
        for (int i = 0; i < quantiles.size(); i++) {
            Double quantile = quantiles.get(i);
            double pos = quantile * n;
            int lowerPos = (int) pos;
            int upperPos = lowerPos + 1;
            T lower = _values.get(lowerPos);
            if (upperPos <= n) {
                T upper = _values.get(upperPos);
                T result = caculator.add(lower, caculator.multiply(caculator.subtract(upper, lower), pos - lowerPos));
                results.put(quantile, result);
            } else
                results.put(quantile, lower);
        }

        return results;
    }

    protected void defensiveReset() {
        if (_values.size() == Integer.MAX_VALUE) {
            _logger.warn("count reached int.max, reset and prevent overflow");

            _values.clear();
        }
    }

}

File: dotnet/src/MyDotey.QuantileEstimator/Classic/ClassicQuantileEstimator.cs
using System;
using System.Collections.Generic;

using NLog;
using MyDotey.Quantile.Value;

/**
 * @author koqizhao
 *
 * Apr 1, 2018
 */
namespace MyDotey.Quantile.Classic
{
    public class ClassicQuantileEstimator<T> : IQuantileEstimator<T>
    {
        private static Logger _logger = LogManager.GetCurrentClassLogger();

        private ClassicQuantileEstimatorConfig<T> _config;

        private volatile List<T> _values;

        public ClassicQuantileEstimator(ClassicQuantileEstimatorConfig<T> config)
        {
            if (config == null)
                throw new ArgumentNullException("config is null");

            _config = config;

            _values = new List<T>();
        }

        public virtual void Add(T value)
        {
            _values.Add(value);
        }

        public virtual Dictionary<Double, T> Get(List<Double> quantiles)
        {
            if (_values.Count == 0)
                return null;

            _values.Sort(_config.Comparer);

            Dictionary<Double, T> results = new Dictionary<Double, T>();
            double n = _values.Count - 1;
            ICalculator<T> caculator = _config.Calculator;
            for (int i = 0; i < quantiles.Count; i++)
            {
                Double quantile = quantiles[i];
                double pos = quantile * n;
                int lowerPos = (int)pos;
                int upperPos = lowerPos + 1;
                T lower = _values[lowerPos];
                if (upperPos <= n)
                {
                    T upper = _values[upperPos];
                    T result = caculator.Add(lower, caculator.Multiply(caculator.Subtract(upper, lower), pos - lowerPos));
                    results[quantile] = result;
                }
                else
                    results[quantile] = lower;
            }

            return results;
        }

        protected virtual void DefensiveReset()
        {
            if (_values.Count == int.MaxValue)
            {
                _logger.Warn("count reached int.max, reset and prevent overflow");

                _values.Clear();
            }
        }
    }
}